// @cco2/carbon-weight/dist/food/types.js
// Data sets available
export var DataE;
(function (DataE) {
    DataE["ADEME_2022"] = "ADEME_2022";
})(DataE || (DataE = {}));
// Food types available
export var FoodE;
(function (FoodE) {
    FoodE["alcohol"] = "alcohol";
    FoodE["bread"] = "bread";
    FoodE["cheese"] = "cheese";
    FoodE["fish"] = "fish";
    FoodE["harvestExotic"] = "harvestExotic";
    FoodE["harvestLocal"] = "harvestLocal";
    FoodE["meatRed"] = "meatRed";
    FoodE["meatWhite"] = "meatWhite";
    FoodE["rice"] = "rice";
    FoodE["soft"] = "soft";
})(FoodE || (FoodE = {}));
// Waste types available
export var WasteE;
(function (WasteE) {
    WasteE["cardboard"] = "cardboard";
    WasteE["fat_butter"] = "fat_butter";
    WasteE["glass_bottle"] = "glass_bottle";
    WasteE["food_brick"] = "food_brick";
    WasteE["milk_bottle"] = "milk_bottle";
    WasteE["none"] = "none";
    WasteE["packaging_glass"] = "packaging_glass";
    WasteE["packaging_pet"] = "packaging_pet";
    WasteE["paper"] = "paper";
    WasteE["plastic_bottle"] = "plastic_bottle";
    WasteE["plastic_film"] = "plastic_film";
    WasteE["tin_can"] = "tin_can";
    WasteE["tray"] = "tray";
    WasteE["tray_ps"] = "tray_ps";
})(WasteE || (WasteE = {}));
