// @cco2/carbon-weight/dist/house/types.js
// Data sets available
export var DataE;
(function (DataE) {
    DataE["ADEME_2022"] = "ADEME_2022";
    DataE["CCO2_2021"] = "CCO2_2021";
    DataE["CCO2_2022"] = "CCO2_2022";
    DataE["CITEPA_2021"] = "CITEPA_2021";
    DataE["CITEPA_2022"] = "CITEPA_2022";
})(DataE || (DataE = {}));
// Climate types available in the country
// Note: Factor introduced and available in CCO2 dataset
export var ClimateE;
(function (ClimateE) {
    ClimateE["H1"] = "H1";
    ClimateE["H2"] = "H2";
    ClimateE["H3"] = "H3"; // Warm
})(ClimateE || (ClimateE = {}));
// Heater Combustible types available
export var HeaterE;
(function (HeaterE) {
    HeaterE["urban"] = "urban";
    HeaterE["electric"] = "electric";
    HeaterE["fuelOil"] = "fuelOil";
    HeaterE["gas"] = "gas";
    HeaterE["coal"] = "coal";
    HeaterE["heatPump"] = "heatPump";
    HeaterE["thermalSolar"] = "thermalSolar";
    HeaterE["propane"] = "propane";
    HeaterE["wood"] = "wood"; // Firewood / Coal - Kg
})(HeaterE || (HeaterE = {}));
// House types available
export var HouseE;
(function (HouseE) {
    HouseE["apartment"] = "apartment";
    HouseE["house"] = "house";
})(HouseE || (HouseE = {}));
// Climate available
export var YearE;
(function (YearE) {
    YearE["old"] = "old";
    YearE["recent"] = "recent"; // After 1975
})(YearE || (YearE = {}));
